File: test/prompt.test.ts
import { describe, expect, it } from "vitest";
import { defaults } from "../src/config";
import { build, languageName, truncate, type PromptInput } from "../src/prompt";

const empty: PromptInput = { stat: "", diff: "", excluded: [] };

describe("languageName", () => {
  it.each([
    ["en", "English"],
    ["ES", "Spanish"],
    [" pt ", "Portuguese"],
    ["Brazilian Portuguese", "Brazilian Portuguese"],
  ])("%j", (input, want) => {
    expect(languageName(input)).toBe(want);
  });
});

describe("build", () => {
  it("includes the contract", () => {
    const out = build(
      { ...defaults(), language: "es", subjectMaxLength: 60 },
      { stat: "a.go | 2 +-", diff: "diff --git a/a.go", excluded: [] },
      "",
    );
    for (const want of [
      "Spanish",
      "60 characters",
      "feat, fix",
      "a.go | 2 +-",
      "diff --git a/a.go",
      "<commit>",
    ]) {
      expect(out).toContain(want);
    }
  });

  it("truncates the diff and says so", () => {
    const out = build(
      { ...defaults(), maxDiffChars: 1000 },
      { ...empty, diff: "Z".repeat(5000) },
      "",
    );
    expect(out).toContain("truncated to 1000 characters");
    expect(out.split("Z").length - 1).toBe(1000);
  });

  it("omits the truncation note when the diff fits", () => {
    expect(build(defaults(), { ...empty, diff: "small" }, "")).not.toContain("truncated");
  });

  it.each([
    ["never", "Do not write one. Subject line only."],
    ["always", "Always write one."],
    ["auto", "only when the change has several distinct parts"],
  ] as const)("body=%s", (mode, want) => {
    expect(build({ ...defaults(), body: mode }, empty, "")).toContain(want);
  });

  it("includes instructions and exclusions", () => {
    const out = build(
      defaults(),
      { ...empty, diff: "d", excluded: ["pnpm-lock.yaml", "a.snap"] },
      "Commit messages must mention the ticket.",
    );
    expect(out).toContain("Project conventions");
    expect(out).toContain("must mention the ticket");
    expect(out).toContain("pnpm-lock.yaml, a.snap");
  });

  it("has no empty conventions section", () => {
    expect(build(defaults(), { ...empty, diff: "d" }, "")).not.toContain("Project conventions");
  });

  it("includes the case and wrap rules", () => {
    const out = build(defaults(), { ...empty, diff: "d" }, "");
    for (const want of [
      "Start the description with a lowercase letter",
      "The scope, when there is one, is lowercase",
      "Wrap every body line at 100 characters",
    ]) {
      expect(out).toContain(want);
    }
  });

  it("omits relaxed rules", () => {
    const out = build(
      { ...defaults(), subjectCase: "any", scopeCase: "any", bodyMaxLineLength: 0 },
      { ...empty, diff: "d" },
      "",
    );
    for (const unwanted of [
      "Start the description with a lowercase letter",
      "The scope, when there is one, is lowercase",
      "Wrap every body line",
    ]) {
      expect(out).not.toContain(unwanted);
    }
  });

  it("truncates the diff by character", () => {
    const out = build(
      { ...defaults(), maxDiffChars: 1000 },
      { ...empty, diff: "ñ".repeat(5000) },
      "",
    );
    expect(out.split("ñ").length - 1).toBe(1000);
  });
});

describe("truncate", () => {
  it.each<[string, string, number, string, boolean]>([
    ["fits exactly", "abc", 3, "abc", false],
    ["cuts ascii", "abcdef", 3, "abc", true],
    ["zero means no limit", "abcdef", 0, "abcdef", false],
    ["empty", "", 3, "", false],
    ["counts characters, not bytes", "ééééé", 3, "ééé", true],
    ["multi-byte that fits by characters but not by bytes", "ééé", 3, "ééé", false],
    ["never splits a character", "aé", 2, "aé", false],
    ["cuts before a multi-byte character", "aéb", 1, "a", true],
    ["never splits a surrogate pair", "a🤖b", 2, "a🤖", true],
  ])("%s", (_, input, max, want, cut) => {
    expect(truncate(input, max)).toEqual([want, cut]);
  });
});

File: internal/prompt/prompt_test.go
package prompt

import (
	"strings"
	"testing"
	"unicode/utf8"

	"github.com/JeanpierreSolis15/commitron/internal/config"
)

func TestLanguageName(t *testing.T) {
	tests := map[string]string{
		"en":                   "English",
		"ES":                   "Spanish",
		" pt ":                 "Portuguese",
		"Brazilian Portuguese": "Brazilian Portuguese",
	}
	for in, want := range tests {
		if got := LanguageName(in); got != want {
			t.Errorf("LanguageName(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestBuildIncludesTheContract(t *testing.T) {
	cfg := config.Defaults()
	cfg.Language = "es"
	cfg.SubjectMaxLength = 60

	out, err := Build(cfg, Input{Stat: "a.go | 2 +-", Diff: "diff --git a/a.go"}, "")
	if err != nil {
		t.Fatal(err)
	}
	for _, want := range []string{"Spanish", "60 characters", "feat, fix", "a.go | 2 +-", "diff --git a/a.go", "<commit>"} {
		if !strings.Contains(out, want) {
			t.Errorf("prompt is missing %q", want)
		}
	}
}

func TestBuildTruncatesTheDiff(t *testing.T) {
	cfg := config.Defaults()
	cfg.MaxDiffChars = 1000
	long := strings.Repeat("Z", 5000)

	out, err := Build(cfg, Input{Diff: long}, "")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(out, "truncated to 1000 characters") {
		t.Error("a truncated diff must say so")
	}
	if n := strings.Count(out, "Z"); n != 1000 {
		t.Errorf("diff was not cut to the budget, got %d characters", n)
	}
}

func TestBuildOmitsTruncationNoteWhenItFits(t *testing.T) {
	out, err := Build(config.Defaults(), Input{Diff: "small"}, "")
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(out, "truncated") {
		t.Error("a diff that fits must not be labelled truncated")
	}
}

func TestBuildBodyModes(t *testing.T) {
	cases := map[string]string{
		"never":  "Do not write one. Subject line only.",
		"always": "Always write one.",
		"auto":   "only when the change has several distinct parts",
	}
	for mode, want := range cases {
		cfg := config.Defaults()
		cfg.Body = mode
		out, err := Build(cfg, Input{}, "")
		if err != nil {
			t.Fatal(err)
		}
		if !strings.Contains(out, want) {
			t.Errorf("body=%q should produce %q", mode, want)
		}
	}
}

func TestBuildIncludesInstructionsAndExclusions(t *testing.T) {
	out, err := Build(
		config.Defaults(),
		Input{Diff: "d", Excluded: []string{"pnpm-lock.yaml", "a.snap"}},
		"Commit messages must mention the ticket.",
	)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(out, "Project conventions") ||
		!strings.Contains(out, "must mention the ticket") {
		t.Error("the instructions section is missing")
	}
	if !strings.Contains(out, "pnpm-lock.yaml, a.snap") {
		t.Error("excluded files should still be named")
	}
}

func TestBuildWithoutInstructionsHasNoEmptySection(t *testing.T) {
	out, err := Build(config.Defaults(), Input{Diff: "d"}, "")
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(out, "Project conventions") {
		t.Error("no instructions means no section header")
	}
}

func TestBuildIncludesCaseAndWrapRules(t *testing.T) {
	out, err := Build(config.Defaults(), Input{Diff: "d"}, "")
	if err != nil {
		t.Fatal(err)
	}
	for _, want := range []string{
		"Start the description with a lowercase letter",
		"The scope, when there is one, is lowercase",
		"Wrap every body line at 100 characters",
	} {
		if !strings.Contains(out, want) {
			t.Errorf("prompt is missing %q", want)
		}
	}
}

func TestBuildOmitsRelaxedRules(t *testing.T) {
	cfg := config.Defaults()
	cfg.SubjectCase = "any"
	cfg.ScopeCase = "any"
	cfg.BodyMaxLineLength = 0

	out, err := Build(cfg, Input{Diff: "d"}, "")
	if err != nil {
		t.Fatal(err)
	}
	for _, unwanted := range []string{
		"Start the description with a lowercase letter",
		"The scope, when there is one, is lowercase",
		"Wrap every body line",
	} {
		if strings.Contains(out, unwanted) {
			t.Errorf("a relaxed config should not produce %q", unwanted)
		}
	}
}

func TestTruncate(t *testing.T) {
	tests := []struct {
		name string
		in   string
		max  int
		want string
		cut  bool
	}{
		{"fits exactly", "abc", 3, "abc", false},
		{"cuts ascii", "abcdef", 3, "abc", true},
		{"zero means no limit", "abcdef", 0, "abcdef", false},
		{"empty", "", 3, "", false},
		{"counts characters, not bytes", "ééééé", 3, "ééé", true},
		{"multi-byte that fits by characters but not by bytes", "ééé", 3, "ééé", false},
		{"never splits a character", "aé", 2, "aé", false},
		{"cuts before a multi-byte character", "aéb", 1, "a", true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, cut := Truncate(tt.in, tt.max)
			if got != tt.want || cut != tt.cut {
				t.Errorf("Truncate(%q, %d) = %q, %v; want %q, %v", tt.in, tt.max, got, cut, tt.want, tt.cut)
			}
			if !utf8.ValidString(got) {
				t.Errorf("Truncate(%q, %d) produced invalid UTF-8", tt.in, tt.max)
			}
		})
	}
}

func TestBuildTruncatesTheDiffByCharacter(t *testing.T) {
	cfg := config.Defaults()
	cfg.MaxDiffChars = 1000
	long := strings.Repeat("ñ", 5000)

	out, err := Build(cfg, Input{Diff: long}, "")
	if err != nil {
		t.Fatal(err)
	}
	if !utf8.ValidString(out) {
		t.Fatal("the prompt contains a half-cut character")
	}
	if n := strings.Count(out, "ñ"); n != 1000 {
		t.Errorf("diff was not cut to 1000 characters, got %d", n)
	}
}
